azure-kusto-ingest/azure/kusto/ingest/ingest_client.py

```diff
@@ -108,3 +127,37 @@ def ingest_from_blob(self, blob_descriptor: BlobDescriptor, ingestion_properties
         content = ingestion_blob_info_json
         queue_client = queue_service.get_queue_client(queue=random_queue.object_name, message_encode_policy=TextBase64EncodePolicy())
         queue_client.send_message(content=content)
+
+    def _validate_endpoint_service_type(self):
+        if self._endpoint_service_type is None or not self._endpoint_service_type.strip():
+            self._endpoint_service_type = self._retrieve_service_type()
+
+        if self._EXPECTED_SERVICE_TYPE != self._endpoint_service_type:
+            has_endpoint = True
+            if self._suggested_endpoint_uri is None or not self._suggested_endpoint_uri.strip():
+                self._suggested_endpoint_uri = self._generate_endpoint_suggestion(self._connection_datasource)
+                if not self._suggested_endpoint_uri.strip():
+                    has_endpoint = False
+            if has_endpoint:
+                raise KustoInvalidEndpointError(self._EXPECTED_SERVICE_TYPE, self._endpoint_service_type, self._suggested_endpoint_uri)
+            else:
+                raise KustoInvalidEndpointError(self._EXPECTED_SERVICE_TYPE, self._endpoint_service_type)
+
+    def _retrieve_service_type(self):
+        return self._resource_manager.retrieve_service_type()
+
+    def _generate_endpoint_suggestion(self, datasource):
+        """The default is not passing a suggestion to the exception String"""
+        endpoint_uri_to_suggest_str = ""
+        if datasource.strip():
+            try:
+                endpoint_uri_to_suggest = urlparse(datasource)  # Standardize URL formatting
+                endpoint_uri_to_suggest = urlparse(endpoint_uri_to_suggest.scheme + "://" + self._INGEST_PREFIX + endpoint_uri_to_suggest.hostname)
+                endpoint_uri_to_suggest_str = endpoint_uri_to_suggest.geturl()
+            except Exception as ex:
+                print(
```

Comment:
I don't think you'd want to print here. I agree we lack of logging mechanism but printing is not the alternative, packages not usually print unless special cases (device auth for example). Just pass here.

Reply:
I agree that it was awkward to print. But I searched the Python SDK and found 1 other instance where we print as a way of providing a warning.

I realize this isn't ideal, but I did it because I couldn't think of a better way, and I think this is better than nothing. Is there a better way of doing it?

Reply:
I don't think this is helpful, but confusing. User gets the actual error, and if the endpoint suggestion failed somehow its a secondary issue. User has nothing to do with this information unlike other places.

Reply:
IMHO, the concept of a warning is a valid and critical one. Sometimes you don't want to fail, but you need to let your user know that something less-than-ideal happened. That's the case here.

If you think the message itself is confusing, we can improve that. But I wrote it like that to provide both some technical data in case they want to dig into the code/debug, and also a high-level suggestion for how they can proceed.

Reply:
I agree with @ArielYehezkely.
As an SDK, stdout and stderr are not yours to write to.

Reply:
I agree with the problem. I had that conundrum when I was writing this. The question is how to make the code as good as it can reasonably be:
1) Ideally, we can add a logging infrastructure for these 2 and any future warnings. The downside to this is that it's out-of-scope for this change and might take time.
2) We already print elsewhere in the Python SDK. Therefore we're already in violation. I suggest we be consistent with our warnings. In that case we can print in both places and add a Work Item to implement #1 in the future.

I think one of these 2 solutions are best.

Reply:
This print is not as important as you make it to be.
If you can't generate the URI and you already throw a less specific exception, just let this warning go.